src/graph.rs

```diff
@@ -56,15 +75,107 @@ impl GitGraph {
 pub struct CommitInfo {
     pub oid: Oid,
     pub branches: Vec<String>,
-    pub branch_traces: Vec<String>,
+    pub branch_trace: Option<String>,
```

Comment:
@adamreichold It would probably be good (or say, better) to use `Option<&str>` here. I guess I would have to store branch names in the `Graph`and mess around with lifetimes (not much experience yet...), but I will give it a try.

Reply:
There are multiple options and `&str` is certainly one but as you already expect, lifetimes are infective to the code base. Alternative to eating the heap allocation would be using a small-buffer optimized string type like [smol_str](https://crates.io/crates/smol_str) or using an arena like [bumpalo](https://crates.io/crates/bumpalo) for storing everything w.r.t. a single lifetime. There are also crates like [easy_string](https://crates.io/crates/easy_strings/) which provide a string type closer to what garbage collected languages offer (which are often [measurably faster than Rust at doing string manipulation](https://users.rust-lang.org/t/rust-vs-go-string-manipulation-performance/36705/45)).

Reply:
Currently trying the `&str` variant, and giving everything a lifetime `'graph`. But now I get this:

```
  --> src\graph.rs:33:13
   |
12 | impl<'graph> GitGraph<'graph> {
   |      ------ lifetime `'graph` defined here
...
30 |         let branches = assign_branches(&repository, &mut commits, indices, settings)?;
   |                                                     ------------ borrow of `commits` occurs here
...
33 |             commits,
   |             ^^^^^^^ move out of `commits` occurs here
...
37 |         Ok(graph)
   |         --------- returning this value requires that `commits` is borrowed for `'graph`
```

The related version of `graph.rs` is [here](https://gist.github.com/mlange-42/96e0a105ffe375d49dc99c446b25e168) (don't want to commit something not working). Maybe this can be solved easily? Will take a break now and try on in 1-2h...

Reply:
I this should be fixable: The issue is borrowing `commits` for `'graph'` which is the whole scope of the function (actually even more it is the lifetime of the return value) and only afterwards moving `commits` into the `graph` struct which could change its address in memory. I think the easiest fix should be to borrow `commits` for a shorter lifetime, more specifically use a signature like
```rust
fn assign_branches(
    repository: &Repository,
    commits: &mut Vec<CommitInfo<'_>>,
    indices: HashMap<Oid, usize>,
    settings: &Settings,
) -> Result<Vec<BranchInfo>, Error> {
``` 
as `BranchInfo` does not contain references and hence you do not need the borrow for that long.

(As an aside, the name `'graph` seems to be a bit of misnomer as it is actually the lifetime of the borrow of `Repository` which as it is being stored within `Graph` necessarily outlasts it. So `'repo` would probably be more fitting.)

But thinking about it a bit more: I expect this run into more fundamental issues soon. You are string to store references in the `branches` field of `Graph` within the `commits` field of the same struct, i.e. to create a self-referential struct which is not easily possible in Rust. While there are crates like [rental](https://crates.io/crates/rental) to help with this, it will probably not make for elegant code.

I see two ways to solve this in a straight-forward manner if you sure that you need to avoid the heap allocation (I am personally not so sure of that): Adopt one of the string types mentioned above to reduce the cost of allocating strings. Or avoid storing string slices and branch names completely. You could just as well replace them by indices into the `branches` vector, e.g.
```
pub struct CommitInfo {
    pub oid: Oid,
    pub branches: Vec<usize>,
    pub branch_trace: Option<usize>,
} 
```
and resolve them by indexing into `graph.branches`. (This could even be more flexible as you do not need to limit yourself to the names but can resolve other fields from `BranchInfo`.)

Reply:
Now that I have the `BranchInfo`s stored in the `Graph`, I also realized that I simply need the indices inside `CommitInfo`, which simplifies all that, just as you wrote.

Originally, I did not think the the heap allocations would be a problem, I merely saw it as a practice for Rust and lifetimes for later occasions where memory usage and/or allocations might be more critical. 😄 

> (As an aside, the name 'graph seems to be a bit of misnomer as it is actually the lifetime of the borrow of Repository which as it is being stored within Graph necessarily outlasts it. So 'repo would probably be more fitting.)

I first created the repo inside `Graph::new()`, therefore the name. I then got a similar error like now forthe commits (IIRC) and thought I could solve it this way.

Thanks for the explanations!